File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int N = nums.length;
        
        int[] left = new int[N];
        int[] right = new int[N];
        
        left[0] = 1;
        right[N - 1] = 1;
        
        // Fill left array
        for (int i = 1; i < N; i++) {
            left[i] = left[i - 1] * nums[i - 1];
        }
        
        // Fill right array
        for (int i = N - 2; i >= 0; i--) {
            right[i] = right[i + 1] * nums[i + 1];
        }
        
        // Another loop
        for (int i = 0; i < N; i++) {
            nums[i] = left[i] * right[i];
        }
        
        return nums;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
from functools import reduce

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prod, zero_count = reduce(lambda a, b: a * (b if b else 1), nums, 1), nums.count(0)
        
        if zero_count > 1:
            return [0] * len(nums)
        else:
            for idx, num in enumerate(nums):
                if zero_count == 1:
                    if num == 0:
                        nums[idx] = prod
                    else:
                        nums[idx] = 0
                else:
                    nums[idx] = prod // num
        return nums